### PR TITLE
New blog post - Implement Stardust

### PR DESCRIPTION
**Pull request by yuanhengzhao**

### CODE REVIEW

File: _posts/henry_zhao/2019-03-06-Implement-Stardust.md [hunk 1]
```diff
@@ -0,0 +1,287 @@
+---
+layout: post
+title: 'Accelerate data visualization with WebGL Platform Stardust'
```
Comment (Fgerthoffert) [inline _posts/henry_zhao/2019-03-06-Implement-Stardust.md]: I would rephrase to "Accelerate data visualization with Stardust, a WebGL Platform"

File: _posts/henry_zhao/2019-03-06-Implement-Stardust.md [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+  title: Software Engineering Blog
+  image: header-logo-crop.png
+  icon: icon-blog
+---
+
+<image src="{{ site.urlimg }}/henry_zhao/future_world.jpg" />
+
+# Introduction
+
+Genomic data is huge and messy. In order to provide researchers a powerful data analysis&management platform, what we present data is the most popular data processing library, d3. It has a wealth of tools that can use svg to portray and manipulate data patterns to the greatest extent possible with different ranges and different forms of pie chart, histograms charts and human figures. For example:
```
Comment (Fgerthoffert) [inline _posts/henry_zhao/2019-03-06-Implement-Stardust.md]: We might want a different word than `messy`, @rosibaj an idea?

File: _posts/henry_zhao/2019-03-06-Implement-Stardust.md [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+
+<image src="{{ site.urlimg }}/henry_zhao/human_body.png" />
+
+However, when the data set is too big(larger than 10000 node), there will an inevitable DOM operation bottleneck. It makes me think about what kind of tools can improve or even replace d3 to be able to visualize data faster.
```
Comment (Fgerthoffert) [inline _posts/henry_zhao/2019-03-06-Implement-Stardust.md]: there will **be** an inevitable DOM

File: _posts/henry_zhao/2019-03-06-Implement-Stardust.md (hunk 2 above)
Comment (Fgerthoffert) [inline _posts/henry_zhao/2019-03-06-Implement-Stardust.md]: First time I went through the paragraph I had some difficulty to figure that the second sentence was referring to d3. Maybe see if you can improve the flow between the two. For example:

> To provide researchers a powerful data analysis&management platform, we traditionally use d3, a popular data processing library. It has a wealth of tools that can use svg to portray and manipulate data patterns.
> For example, GDC's human figure was implemented in d3:

File: _posts/henry_zhao/2019-03-06-Implement-Stardust.md [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+So first we should know what is svg and why it has bottleneck.
```
Comment (Fgerthoffert) [inline _posts/henry_zhao/2019-03-06-Implement-Stardust.md]: `So first we should understand what SVG is, and what are the possible causes of a bottleneck.`

File: _posts/henry_zhao/2019-03-06-Implement-Stardust.md [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+
+### SVG
+
+SVG is a vector graphic format—based on XML and is used to display a variety of graphics on the Web and other environments. Since SVG are defined in XML, every SVG element is appended to the DOM which is big cost(It is the reason that reactjs use virtual DOM to avoid DOM operation).
```
Comment (Fgerthoffert) [inline _posts/henry_zhao/2019-03-06-Implement-Stardust.md]: ...format, based on XML, and is...

Comment (Fgerthoffert) [inline _posts/henry_zhao/2019-03-06-Implement-Stardust.md]: `which is big cost` => `which has a big performance cost`

File: _posts/henry_zhao/2019-03-06-Implement-Stardust.md [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+
+### WebGL
+
+There is a potential way to replace svg which is WebGl. WebGL (Web Graphics Library) is a JavaScript API for rendering interactive 3D and 2D graphics within any compatible web browser without the use of plug-ins. WebGL does so by introducing an API that closely conforms to OpenGL ES2.0 that can be used in HTML5 canvas elements. It provides local computing resources API to accelerate the rendering of the graphics and avoids massive operations of DOM. Therefore, it is naturally suitable for large data manipulation.
+
+[WebGL Example](https://webglsamples.org/)
+
+<image src="{{ site.urlimg }}/henry_zhao/webgl_building.gif" />
+
+There are a lot of WebGL libraries for data visualization. I will use stardust[stardust](https://stardustjs.github.io/) as an example to demonstrate how to implement it.
```
Comment (Fgerthoffert) [inline _posts/henry_zhao/2019-03-06-Implement-Stardust.md]: I would put "log of WebGL libraries" as a external link listing those, if such page exists.

File: _posts/henry_zhao/2019-03-06-Implement-Stardust.md [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+
+### Stardust
+
+First install stardust:
```
Comment (Fgerthoffert) [inline _posts/henry_zhao/2019-03-06-Implement-Stardust.md:50]: Are you missing npm install ? The command below is importing stardust, not installing.

File: _posts/henry_zhao/2019-03-06-Implement-Stardust.md [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+          1000
+        );
+        t1.check(e);
+      } else {
+        circles2.data([]);
+        circles2.render();
+      }
+    };
+```
+
```
Comment (Fgerthoffert) [inline _posts/henry_zhao/2019-03-06-Implement-Stardust.md:151]: Add a small conclusion explaining what is obtained along with a link to a codepen or similar with the actual "live"example.

File: _posts/henry_zhao/2019-03-06-Implement-Stardust.md [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+      this._canvasNode.width,
+      this._canvasNode.height
+    );
+    this._marks.rects.render();
+    this._marks.circles.render();
+    this._platform.endPicking();
+  }
+}
+```
+
```
Comment (Fgerthoffert) [inline _posts/henry_zhao/2019-03-06-Implement-Stardust.md:175]: Same than previous, add working code in a codepen or similar

File: _posts/henry_zhao/2019-03-06-Implement-Stardust.md [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+### Conclusion
+
+Although my implementation doesn't cover many real scenarios, there is no doubt that stardust has a better rendering performance than d3. A fully replacement of d3 will never happen in a real world due to the all the powerful components d3 provided and the advantages of svg in event handling. D3 are also a good complementary for Stardust(WebGL). A mix of the two in applications is recommended by the creator of Stardust: d3 being used to render a scatterplot’s axes and handle interactions such as range selections, but Stardust used to render and animate the points.
```
Comment (Fgerthoffert) [inline _posts/henry_zhao/2019-03-06-Implement-Stardust.md]: If you rephrase a bit your second sentence to actually introduce your last point earlier. For example:
`Nevertheless, Stardust doesn't aim at replacing d3, but more at improving performance in rendering and animating points. The creator of Stardust actually recommends....`

**Review by Fgerthoffert — APPROVED** [review]
Great article!

File: _posts/henry_zhao/2019-03-06-Implement-Stardust.md (hunk 2 above)
Comment (rosibaj) [reply _posts/henry_zhao/2019-03-06-Implement-Stardust.md]: `complicated` ?

Comment (rosibaj) [inline _posts/henry_zhao/2019-03-06-Implement-Stardust.md]: analysis & management <-- need a space here

Comment (rosibaj) [reply _posts/henry_zhao/2019-03-06-Implement-Stardust.md]: `pie charts`, <-- just needs an s too

File: _posts/henry_zhao/2019-03-06-Implement-Stardust.md (hunk 3 above)
Comment (rosibaj) [reply _posts/henry_zhao/2019-03-06-Implement-Stardust.md]: `10000 nodes` <--with an s here as well

File: _posts/henry_zhao/2019-03-06-Implement-Stardust.md (hunk 5 above)
Comment (rosibaj) [inline _posts/henry_zhao/2019-03-06-Implement-Stardust.md]: `(It is the reason that reactjs use virtual DOM to avoid DOM operation).` . I would go ahead and make this a full sentence not () as its valuable information. 
Ex:
`This performance issue is the reason that reactjs use virtual DOM to avoid DOM operations.`

File: _posts/henry_zhao/2019-03-06-Implement-Stardust.md (hunk 6 above)
Comment (rosibaj) [reply _posts/henry_zhao/2019-03-06-Implement-Stardust.md]: consistently capitolize `svg` to `SVG` here, as well as `WebGl` to `WebGL`.

Comment (rosibaj) [inline _posts/henry_zhao/2019-03-06-Implement-Stardust.md]: `stardust[stardust]` --> `[stardust] `

File: _posts/henry_zhao/2019-03-06-Implement-Stardust.md [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+        circles2.render();
+      }
+    };
+```
+
+In general, Stardust has all the basic functionality of d3.
+
+### Performance
+
+To demonstrate the performance of stardust, I will mock up oncogrid features in GDC portal. The purpose of oncogrid is to demonstrate the mutation pattern between genes and patients. In the gif, There are less than 10,000 data points in the grid.
```
Comment (rosibaj) [inline _posts/henry_zhao/2019-03-06-Implement-Stardust.md]: `In the gif, There` -->` In the gif, there`

**Review by rosibaj — CHANGES_REQUESTED** [review]
Good read - great topic and discussion.  Just a few small edits!

File: _posts/henry_zhao/2019-03-06-Implement-Stardust.md (hunk 2 above)
Comment (yuanhengzhao) [reply _posts/henry_zhao/2019-03-06-Implement-Stardust.md]: disorganized? @Fgerthoffert @rosibaj

**Comment by yuanhengzhao** [comment]
@Fgerthoffert @rosibaj  Made all the changes.

**Comment by b-f-chan** [comment]
Completed for: #122
